pkg/api/customization/monitor/metric_action.go

```diff
@@ -119,53 +122,75 @@ func (h *MetricHandler) Action(actionName string, action *types.Action, apiConte
 		}
 		apiContext.Response.Write(res)
 
-	case listclustermetricname, listprojectmetricname:
+	case listclustermetricname:
+		var input v3.ClusterMetricNamesInput
+		actionInput, err := parse.ReadBody(apiContext.Request)
+		if err != nil {
+			return err
+		}
+		if err = convert.ToObj(actionInput, &input); err != nil {
+			return err
+		}
 
-		var clusterName, projectName, appName, saNamespace string
-		var err error
+		clusterName := input.ClusterName
+		if clusterName == "" {
+			return fmt.Errorf("clusterName is empty")
+		}
 
-		if actionName == listclustermetricname {
-			var input v3.ClusterMetricNamesInput
-			actionInput, err := parse.ReadBody(apiContext.Request)
-			if err != nil {
-				return err
-			}
-			if err = convert.ToObj(actionInput, &input); err != nil {
-				return err
-			}
+		userContext, err := h.clustermanager.UserContext(clusterName)
+		if err != nil {
+			return fmt.Errorf("get usercontext failed, %v", err)
+		}
 
-			clusterName = input.ClusterName
-			if clusterName == "" {
-				return fmt.Errorf("clusterName is empty")
-			}
+		appName, saNamespace := monitorutil.ClusterMonitoringInfo()
+		token, err := getAuthToken(userContext, appName, saNamespace)
```

Comment:
can you explain what this token is, why we are getting it, and how we use it?
I just want to make sure we arent doing anything bad from a security perspective

Reply:
This token is created when enable cluster/project level monitoring, when call prometheus api will add this token, then prometheus auth component will return the auth result. There still have problem, because this token just have two level(cluster/project), can't really match the rbac in rancher. For example, use project level token to query project prometheus will allow all data that have namespace label under that project, for a project user who can't access workload but can access service, should only allow them to execute service relate expression(or execute workload expression but return no result), but now don't have a good solution for match expression(prometheus data) with rancher rbac. @thxCode @orangedeng How do you think from a security perspective?